extension/src/state_aggregate.rs

```diff
@@ -0,0 +1,342 @@
+//! SELECT duration_in('STOPPED', states) as run_time, duration_in('ERROR', states) as error_time FROM (
+//!   SELECT state_agg(time, state) as states FROM ...
+//! );
+//!
+//! Currently requires loading all data into memory in order to sort it by time.
+
+#![allow(non_camel_case_types)]
+
+use pgx::*;
+
+use aggregate_builder::aggregate;
+use flat_serialize::*;
+use state_aggregate::DurationInState;
+
+use crate::{
+    flatten,
+    palloc::{Inner, Internal},
+    pg_type,
+    raw::{bytea, TimestampTz},
+    ron_inout_funcs,
+};
+
+#[aggregate]
+impl state_agg {
+    type State = state_aggregate::TransState;
+
+    const PARALLEL_SAFE: bool = true;
+
+    fn transition(
+        state: Option<State>,
+        #[sql_type("timestamptz")] ts: TimestampTz,
+        #[sql_type("text")] value: Option<String>,
+    ) -> Option<State> {
+        let value = match value {
+            None => return state,
+            Some(value) => value,
+        };
+        let mut state = match state {
+            None => state_aggregate::TransState::new().into(),
```

Comment:
I don't think this `into` is needed.  I believe clippy will complain about this (check with `cargo clippy --features 'pg13 pg_test'`).